Modularise themes state.

sgomes opened the pull request:
This PR is one of many working on modularising state in Calypso. This one handles themes.

For mode details on state modularisation, see #39261 and p4TIVU-9lM-p2.

In the interest of keeping this PR short, I will create a subsequent PR for moving theme-related selectors under `state/themes`, instead of handling those changes here too.

#### Changes proposed in this Pull Request

* Modularise themes state
* Create a few selectors to replace direct state access
* Minor lint fixes to touched files

#### Testing instructions

* Install [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en).
* Open DevTools and switch to the Redux tab.
* Open `/read`.
* Click `State` on the top right to ensure you see all state, and not just the diff for the last action: 
![image](https://user-images.githubusercontent.com/409615/73774406-d3e87d80-477b-11ea-9f59-3e42cc00101f.png)
* Analyse the tree and note that there's no `themes` key, even if you expand the list of keys. This indicates that the themes state hasn't been loaded yet.
* Click `My Sites` on the masterbar.
* Verify that a `themes` key is added with the themes state. This indicates that the themes state has now been loaded.
* Verify that theme-related functionality works normally. This indicates that I didn't mess up.

sgomes committed:
Modularise themes state.

sgomes commented:
Waiting for the ICFY results, but this PR should save ~2KB (compressed) on `entry-main`.

sgomes committed:
Update tests to work with modularised state.

tyxla reviewed (changes requested):
This seems to work well when tested with the steps you mentioned on the client side, but the SSR seems to be broken.

When viewing locally in incognito (not logged in) http://calypso.localhost:3000/themes I first see this:

![](https://cldup.com/XRwAHCqRd1.png)

and then I see this broken view:

![](https://cldup.com/nOiRMaaU2t.png)

while on production it works well:

![](https://cldup.com/XPZjVb4DT6.png)

The same applies for the single theme page:

![](https://cldup.com/htIC76EkV6.png)

where on production it looks like this:

![](https://cldup.com/_abEm6cSiB.png)

sgomes commented:
> This seems to work well when tested with the steps you mentioned on the client side, but the SSR seems to be broken.

Excellent catch, @tyxla! It looks like modularised state isn't working properly on the server side, and themes would be the first piece of modularised state that's supposed to work on the server. I'm preparing the necessary changes.

sgomes committed:
Fix modularised state on the server.

The server registers a new store for each request, so some of the
assumptions we were making on the client side are no longer true.

sgomes commented:
@tyxla SSR should be working with the latest commit.

noahtallen reviewed (approved):
I tested several things on the page, and this seems to be working as expected! 😁 @Addison-Stavlo might have more edge cases worth checking.

- Activating a recommended theme gives you an option to go to the block editor
- Activating an older theme gives you an option to customize
- The logged out themes page appears to work correctly now
- Filters appear to be working correctly
- Activating themes does indeed activate the theme
- Redux `themes` state is only visible once you go to "my sites" (any reason why we don't load this even later; like when visiting the themes page?)

Reader view:
<img width="538" alt="Screen Shot 2020-02-13 at 2 47 25 PM" src="https://user-images.githubusercontent.com/6265975/74485990-e136f380-4e70-11ea-8d25-f73067b2b21c.png">

After visiting "my sites"
<img width="553" alt="Screen Shot 2020-02-13 at 2 48 06 PM" src="https://user-images.githubusercontent.com/6265975/74485982-d8462200-4e70-11ea-9fc0-f322a9cba4cd.png">

tyxla reviewed (approved):
LGTM, nice work! 🚢  Thanks @sgomes!

sgomes commented:
> * Redux `themes` state is only visible once you go to "my sites" (any reason why we don't load this even later; like when visiting the themes page?)

Do you mean when visiting `/themes`? The state is there for me when I open it:

![image](https://user-images.githubusercontent.com/409615/74524682-7096e200-4f17-11ea-801b-d9656f491d21.png)

sgomes commented:
Thank you so much for the thorough reviews, @noahtallen and @tyxla! 🙇

sgomes commented:
Leaving this merge for Monday, since it involves changes to the server side.

noahtallen commented:
> Do you mean when visiting /themes? The state is there for me when I open it:

No, I meant that the `themes` state is loaded as soon as you open "My Sites," but the state isn't completely used until you go to `/themes`. So I was wondering if it could be loaded even later, when accessing `/themes`, rather than immediately on going to "My Sites".

Now that I think about it, there are probably other places in "my sites" that use the theme state, so it probably doesn't make sense to load it even later.

sgomes commented:
> No, I meant that the `themes` state is loaded as soon as you open "My Sites," but the state isn't completely used until you go to `/themes`. So I was wondering if it could be loaded even later, when accessing `/themes`, rather than immediately on going to "My Sites".

Oh, I see! Sorry that I totally misunderstood you there 🙂 

I'm not entirely sure what's loading themes when navigating to "My Sites". The new mechanism works automatically through the dependency graph, so all we know is that there's something in "My Sites" that's loading a theme-related selector or action. As modularisation moves forward, it will be interesting to look for these "early loaders" and see if we can delay things further!
